42/README.md

```diff
@@ -0,0 +1,71 @@
+
+Currently, there is no active strategy to address breaking changes in BigchainDB 2.0 i.e. there is no general approach to migration such that breaking changes in transaction specs can be handled. This document is an attempt in quantifying scenarios wherein a breaking might occur and propose a generalized strategy to handle these situations.
+
+## Hard fork scenarios
+1. Change in any of the transaction specs
+    - Change in signature algorithm
+    - Change in hashing functions
+    - Change in schema
+2. Change in storage schema
+
+## Handling Hard Forks
+
+### Type 1 scenarios
+In case of change to spec it might not be possible to port all existing validated transaction to the new format. In such cases one needs to make a distinction between transactions which have already been validated and stored in the Blockchain from new incoming transactions. Furthermore, a suitable trigger is needed so that the network can collectively jump to the new logic else there could be situations wherein the network might not be able to commit any new blocks because of honest nodes running different validation logic. So, it is apparent that a mechanism for transitioning to this new validation logic is need. Below is one such procedure to achieve the same,
+
+- In order to decide when the validation logic should be upgraded to its newer version a TEP is initiated which defines the future block `height` at which to upgrade and the corresponding `BigchainDB` class to use.
+
+   NOTE: It is assumed that the new validation logic is implemented by extending the existing `BigchainDB` class and at the time of starting BigchainDB an instance of each of these classes is passed to the ABCI `App` i.e.,
+   ```python
+   validation = {
+       "v1": {
+           "start_block_height": 0,
+           "validation_class": "BigchainDB",
+       }
+       "v2": {
+           "start_block_height": 20940394,
+           "validation_class": "BigchainDBv2",
+       }
+       ...
+   }
+   ```
+
+- Each node operator needs to approve the TEP proposal in order for the validation logic to be upgraded.
+
+- If a super-majority of the network i.e. `> 2/3` voting power agrees with the change and the proposed height is still lies in the future then the proposed TEP is commit-ed and a new validation version is added (see above)
+
+NOTE: There can only be one-and-only one ongoing hard fork i.e. concurrent hard forks should not be not allowed.
+
+
+### Type 2 scenarios
+
+These scenarios don't actually impact the sanctity of the blockchain but rather the shape of the data. In order to handle such scenarios following two approaches can used,
+
+#### Approach 1
+
+Create migration scripts for each new upgrade i.e. with each new release which results in a change in the shape of the data we create corresponding migrations scripts. Needless to say that writing such scripts might be tedious and in the worst case users can only incrementally upgrade to newer version
+
+#### Approach 2
```

Comment:
This sounds like a decent approach. but some sort of basic upgrade prototype should be made before, we merge this BEP. I am quite unsure about how Tendermint might behave at run time changes for validator/followers in `config.toml`

Reply:
> but some sort of basic upgrade prototype should be made before

Could you please elaborate?

> I am quite unsure about how Tendermint might behave at run time changes for validator/followers in `config.toml`

The running validator's `config.toml` is not changed, rather we create a follower node which we configure to talk to this validator node. Once the follower node is up to pace with the validator node we change the `proxy_app` of the validator node to the BigcahinDB server connected the follower node.